File: app/src/main/java/com/denproj/posmanongjaks/viewModel/SettingsFragmentViewmodel.java
package com.denproj.posmanongjaks.viewModel;

import androidx.lifecycle.ViewModel;

import com.denproj.posmanongjaks.util.AsyncRunner;
import com.denproj.posmanongjaks.util.OnDataReceived;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class SettingsFragmentViewmodel extends ViewModel {

    FirebaseAuth firebaseAuth = FirebaseAuth.getInstance();

    @Inject
    public SettingsFragmentViewmodel() {
    }

    public void resetPassword(boolean isConnectionReachable, OnPasswordResetFinished onPasswordResetFinished) {

        FirebaseUser firebaseUser = firebaseAuth.getCurrentUser();
        if (isConnectionReachable && isUserSignedIn() && firebaseUser.getEmail() != null) {
            String email = firebaseUser.getEmail();
            firebaseAuth.sendPasswordResetEmail(email).addOnCompleteListener(task -> {
                if (task.isSuccessful()) {
                    firebaseAuth.signOut();
                }
            });
            firebaseAuth.addAuthStateListener(firebaseAuth -> {
                if (firebaseAuth.getCurrentUser() == null) {
                    onPasswordResetFinished.onSuccess();
                }
            });
            return;
        }

        onPasswordResetFinished.onFail();
    }

    public void clearSavedLoginAndLogout(OnLogOutSuccessful onLogOutSuccessful) {
        AsyncRunner.runAsync(new AsyncRunner.MiniRunner<Void>() {
            @Override
            public Void onBackground() throws Exception {
                return null;
            }

            @Override
            public void onUI(Void result) {
                deleteFirebaseUser(new OnDataReceived<Void>() {
                    @Override
                    public void onSuccess(Void result) {
                        onLogOutSuccessful.onSuccess();
                    }

                    @Override
                    public void onFail(Exception e) {
                        onLogOutSuccessful.onFail(e);
                    }
                });
            }

            @Override
            public void onError(Exception e) {
                onLogOutSuccessful.onFail(e);
            }
        });
    }

    private boolean isUserSignedIn() {
        return firebaseAuth.getCurrentUser() != null;
    }

    private void deleteFirebaseUser(OnDataReceived<Void> onDataReceived) {
        if (isUserSignedIn()) {
            firebaseAuth.getCurrentUser().delete().addOnCompleteListener(task -> {
                if (task.isSuccessful()) {
                    onDataReceived.onSuccess(null);
                } else {
                    onDataReceived.onFail(task.getException());
                }
            });
        }
    }

    public interface OnPasswordResetFinished {
        void onSuccess();
        void onFail();
    }

    public interface OnLogOutSuccessful {
        void onSuccess();
        void onFail(Exception e);
    }

}

File: app/src/main/java/com/denproj/posmanongjaks/viewModel/SettingsFragmentViewmodel.kt
package com.denproj.posmanongjaks.viewModel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.denproj.posmanongjaks.repository.impl.SessionRepository
import com.denproj.posmanongjaks.util.NetworkStatusMonitor
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class SettingsFragmentViewmodel @Inject constructor(val sessionRepository: SessionRepository) : ViewModel() {
    var firebaseAuth: FirebaseAuth = FirebaseAuth.getInstance()
    val mutableLiveData = MutableLiveData<Boolean>()
    val networkStatusMonitor = NetworkStatusMonitor()

    init {
        viewModelScope.launch {
            networkStatusMonitor.isConnected.collect {
                mutableLiveData.value = it
            }
        }
    }

    fun resetPassword(
        isConnectionReachable: Boolean,
        onPasswordResetFinished: OnPasswordResetFinished
    ) {
        val firebaseUser = firebaseAuth.currentUser
        if (isConnectionReachable && isUserSignedIn && firebaseUser!!.email != null) {
            val email = firebaseUser.email
            firebaseAuth.sendPasswordResetEmail(email!!)
                .addOnCompleteListener { task: Task<Void?> ->
                    if (task.isSuccessful) {
                        firebaseAuth.signOut()
                    }
                }
            firebaseAuth.addAuthStateListener { firebaseAuth: FirebaseAuth ->
                if (firebaseAuth.currentUser == null) {
                    onPasswordResetFinished.onSuccess()
                }
            }
            return
        }

        onPasswordResetFinished.onFail()
    }

    fun logout(onLogOutSuccessful: OnLogOutSuccessful) {
        if (isUserSignedIn) {
            viewModelScope.launch {
                firebaseAuth.signOut()
                sessionRepository.clearSession()
                onLogOutSuccessful.onSuccess()
            }
        } else {
            viewModelScope.launch {
                sessionRepository.clearSession()
                onLogOutSuccessful.onSuccess()
            }
            onLogOutSuccessful.onFail(Exception("User already signed out."))
        }
    }


    private val isUserSignedIn: Boolean
        get() = firebaseAuth.currentUser != null

    interface OnPasswordResetFinished {
        fun onSuccess()
        fun onFail()
    }

    interface OnLogOutSuccessful {
        fun onSuccess()
        fun onFail(e: Exception?)
    }


    interface UserTask {
        fun success()
        fun failed(e: Exception?)
    }
}
